### PR TITLE
Fix for issue #2690 - Authenticated support form

### PR DESCRIPTION
**Pull request by AlexBHarley**
Fix for issue #2690 . This is an initial proof of concept PR. Lot's will have to change - just looking for feedback etc and to see if this is something you're interested in.

### CODE REVIEW

**Comment by maartenba** [comment]
To me the currently committed code looks good. What changes are you planning to do after this one?

Adding @xavierdecoster as he's working on support forms in general.

**Comment by AlexBHarley** [comment]
I was using the ReportPackageRequest.cs (RPR) as a guide for ContactSupportRequest.cs (CSR). You'll see that the RPR file has a FillIn function that acts as a mapper for the bodyTemplate in MessageService.cs. The CSR file doesn't have this and I'm just using a string.Format in the SendSupportContactEmail function in MessageService.cs. Was just thinking this is inconsistent. However if you're happy to merge I won't complain

**Comment by AlexBHarley** [comment]
@maartenba @xavierdecoster is this good to be merged?

File: src/NuGetGallery/Services/ContactSupportRequest.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mail;
+using System.Web;
```
Comment (xavierdecoster) [inline src/NuGetGallery/Services/ContactSupportRequest.cs]: nit: please remove unused usings and sort

File: src/NuGetGallery/Controllers/PagesController.cs [hunk 2]
```diff
@@ -39,7 +44,42 @@ public virtual ActionResult About()
 
         public virtual ActionResult Contact()
         {
-            return View();
+            var model = new ContactSupportViewModel();
+
+            if (Request.IsAuthenticated)
+            {
+                model.ConfirmedUser = true;
```
Comment (xavierdecoster) [inline src/NuGetGallery/Controllers/PagesController.cs]: `ConfirmedUser` seems like an overloaded term here, as the user is just authenticated (and not necessarily e.g. a user that confirmed his email address). Perhaps rename to `AuthenticatedUser`, or just use `User.Identity.IsAuthenticated` in the view?

File: src/NuGetGallery/ViewModels/ContactSupportViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using System.Linq;
+using System.Web;
+using System.Web.Mvc;
+
+namespace NuGetGallery.ViewModels
+{
+    public class ContactSupportViewModel
+    {
+        [Required(ErrorMessage ="Please enter a subject line.")]
+        [StringLength(100)] //This is something I'm not sure on
```
Comment (xavierdecoster) [inline src/NuGetGallery/ViewModels/ContactSupportViewModel.cs]: Sounds reasonable to limit the subject to 100 characters.

File: src/NuGetGallery/ViewModels/ContactSupportViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    public class ContactSupportViewModel
+    {
+        [Required(ErrorMessage ="Please enter a subject line.")]
+        [StringLength(100)] //This is something I'm not sure on 
+        // - it comes down to what is best for Nuget/Microsoft support
+        [Display(Name ="Subject Line")]
+        public string SubjectLine { get; set; }
+
+        [Required(ErrorMessage ="Please enter a message.")]
+        [AllowHtml]
```
Comment (xavierdecoster) [inline src/NuGetGallery/ViewModels/ContactSupportViewModel.cs]: Not convinced we should allow HTML input here.

File: src/NuGetGallery/Views/Pages/Contact.cshtml [hunk 5]
```diff
@@ -45,6 +47,35 @@
...
+    @if (Model.ConfirmedUser)
+    {
+        <h3>The above doesn't help? Get in touch with us using the form below</h3>
+        <p>
+            A member of our Support team will be in touch with you soon.
+        </p>
+        using (Html.BeginForm())
+        {
+            <div class="form-field">
+                @Html.LabelFor(m => m.SubjectLine, "Subject line")
```
Comment (xavierdecoster) [inline src/NuGetGallery/Views/Pages/Contact.cshtml]: No need for the second parameter here as you defined the `Display` attribute on the model's property.

File: src/NuGetGallery/Views/Pages/Contact.cshtml [hunk 6]
```diff
@@ -45,6 +47,35 @@
...
+                @Html.TextAreaFor(m => m.SubjectLine, 1, 50, null)
+            </div>
+            <div class="form-field">
+                @Html.LabelFor(m => m.Message, "Message for support")
```
Comment (xavierdecoster) [inline src/NuGetGallery/Views/Pages/Contact.cshtml]: Same comment as above. Update the `Display` attribute instead.

File: src/NuGetGallery/Content/PageStylings.css [hunk 7]
```diff
@@ -405,6 +405,10 @@ form .async-upload-progress-advance {
     top: 4.6em;
 }
 
+#contact-support-message {
```
Comment (xavierdecoster) [inline src/NuGetGallery/Content/PageStylings.css]: Update line 404 to `#report-abuse-message, #contact-support-message {` as they share the same style.

File: src/NuGetGallery/Controllers/PagesController.cs [hunk 8]
```diff
@@ -1,9 +1,12 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using NuGetGallery.Services;
+using NuGetGallery.ViewModels;
 using System;
 using System.Linq;
 using System.Net;
+using System.Net.Mail;
```
Comment (xavierdecoster) [inline src/NuGetGallery/Controllers/PagesController.cs]: Remove unused usings and sort.

File: src/NuGetGallery/Controllers/PagesController.cs [hunk 9]
```diff
@@ -39,7 +44,42 @@ public virtual ActionResult About()
...
+            var model = new ContactSupportViewModel();
+
+            if (Request.IsAuthenticated)
+            {
+                model.ConfirmedUser = true;
+            }
+            return View(model);
+        }
+
+        [HttpPost]
```
Comment (xavierdecoster) [inline src/NuGetGallery/Controllers/PagesController.cs]: Add the `[Authorize]` attribute on this action.

File: src/NuGetGallery/Controllers/PagesController.cs [hunk 10]
```diff
@@ -39,7 +44,42 @@ public virtual ActionResult About()
...
+        public virtual ActionResult Contact(ContactSupportViewModel contactForm)
+        {
+            if (!ModelState.IsValid)
+            {
+                return View();
+            }
+
+            User user = null;
+            if (Request.IsAuthenticated)
```
Comment (xavierdecoster) [inline src/NuGetGallery/Controllers/PagesController.cs]: The `Authorize` attribute ensures only authenticated requests reach this code path.

**Comment by xavierdecoster** [comment]
@AlexBHarley Thanks for this contribution! Looks ok to me. If you can address the feedback and squash your commits, I can take it from there. (have a parallel change/work in progress that relates to this)

File: src/NuGetGallery/Views/Pages/Contact.cshtml [hunk 11]
```diff
@@ -45,6 +47,35 @@
     <p>
         Need help with your NuGet.org account? Please feel free to <span id="emailUs"></span>
     </p>
+
+    @if (Model.ConfirmedUser)
+    {
+        <h3>The above doesn't help? Get in touch with us using the form below</h3>
```
Comment (xavierdecoster) [inline src/NuGetGallery/Views/Pages/Contact.cshtml]: I think the paragraph on line 48 can be moved after line 77, inside the unauthenticated block.
If for some reason people can't login any longer, they can send us an email. If they can login, I'd prefer the form is used instead.

**Comment by AlexBHarley** [comment]
@xavierdecoster Thanks. Have made the changes and squashed commits. Does it look okay now?

**Comment by xavierdecoster** [comment]
Thanks! Merged with https://github.com/NuGet/NuGetGallery/commit/382c824d87db7939260d8f0e6a0f4a70f903e0a6.

Should go live with deployment https://github.com/NuGet/NuGetGallery/issues/2859

**Pull request closed without merging**